### ISSUE TITLE
Support for Early Hints (103) ?

### ISSUE THREAD

**Issue by jasonwilliams** [fastify/fastify#2683]
This is more of a question than a request to implement a feature.

How possible is it to send a 103 response with a Link header before sending the main response back to the browser?

From what I understand once a response sends something the state of that response is “sent”.

## Motivation

HTTP/2 Server Push has been killed off and removed from Chrome. https://groups.google.com/a/chromium.org/g/blink-dev/c/K3rYLvmQUBY/m/vOWBKZGoAQAJ

The focus instead is on Early Hints, which allows the server to send a preliminary response back to the browser which has a Link header followed by what resources the browser can download while it waits. Then followed by the final response.

This isn’t supported fully in browsers yet but there’s experiments happening to see if it’s worth pursuing.

Is this something frameworks can implement or does there need to be changes at Node level?

Spec:
https://httpwg.org/specs/rfc8297.html

https://chromium.googlesource.com/chromium/src/+/master/docs/early-hints.md

https://www.fastly.com/blog/beyond-server-push-experimenting-with-the-103-early-hints-status-code

**Comment by mcollina** [comment]
You could write directly to `reply.raw.socket` an HTTP response with 103 status code.

**Comment by jasonwilliams** [comment]
@mcollina thanks I was not aware of that!

**Comment by zekth** [comment]
like so?
```javascript
fastify.get("/", async (request, reply) => {
  const links = [
    "Link: </style.css>; rel=preload; as=style",
    "Link: </script.js>; rel=preload; as=script",
  ];
  const rawHttpRequest = `HTTP/1.1 103 Early Hints\r\n${links.join(
    "\r\n"
  )}\r\n\r\n`;
  reply.raw.socket.write(rawHttpRequest);

  return { hello: "world" };
});
```

returns
```
curl -D - http://localhost:3000    

HTTP/1.1 103 Early Hints
Link: </style.css>; rel=preload; as=style
Link: </script.js>; rel=preload; as=script

HTTP/1.1 200 OK
content-type: application/json; charset=utf-8
content-length: 17
Date: Thu, 12 Nov 2020 14:18:21 GMT
Connection: keep-alive

{"hello":"world"}
```

**Comment by mcollina** [comment]
That'd be the idea, yes

**Comment by zekth** [comment]
it's an interesting feature and we can make a plugin for that. I'll make a draft repo for it.

**Comment by zekth** [comment]
Here is the draft repo i've written: https://github.com/zekth/fastify-early-hints
Feel free to comment / contribute.
Also for testing, it seems that i can't acces the full HTTP message throught the `reply.raw`, i might need to spawn a `curl` process to check the message. Any other idea?

**Comment by delvedor** [comment]
I love so much the Fastify plugin system ❤️

**Comment by mcollina** [comment]
Amazing work @zekth! Is it on npm yet? PR to the ecosystem page?

**Comment by Eomm** [comment]
> Any other idea?

I think 103 status is not supported yet by the node core (cause it is still in draft IMHO).
For example there the `100-Continue` is supported by the [`http` client](https://nodejs.org/api/http.html#http_event_continue) but 103 doesn't emit any events and [searching on the nodejs repo](https://github.com/nodejs/node/search?q=HTTP_STATUS_EARLY_HINTS) there isn't such test.

So spawn+curl seems the only way to check it right now.

NB: with http2 the plugin trigger an error, should be a warning be emitted that tell the user to switch to `server push` in this scenario?

**Comment by zekth** [comment]
@mcollina currently i'm facing some issues with the write operation to the response or to the socket. For example we don't wait the end of the write operation to close the response with fastify, leading to errors. I'm fixing this. Once everything is fixed with the first draft and tests are implemented a first release on npm is thinkable yes.

@Eomm will have to dig on http2 yes.

**Comment by zekth** [comment]
Just a heads-up, this is now supported as an experimental feature in Chromium 95: https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/early-hints.md

**Comment by derhuerst** [comment]
FYI: [On 2022-06-21, Chrome 103 became stable](https://chromereleases.googleblog.com/2022/06/stable-channel-update-for-desktop_21.html), which [supports `103`](https://developer.chrome.com/blog/new-in-chrome-103/#http103)

**Comment by mcollina** [comment]
Support for early hints landed in Node.js core: https://github.com/nodejs/node/pull/44180.
This will likely be in Node.js v18.8.0 and likely backported to Node.js v16.
I'm not sure we should do _anything_ to support it (possibly we should forward the call to Node core).

Something that would be interesting is to build a plugin that automatically inspect responses and automatically collect metadata and send them through on next call. Take a look at https://github.com/google/node-h2-auto-push for inspiration.

**Comment by Eomm** [comment]
There is this plugin https://github.com/zekth/fastify-early-hints from @zekth we could start from.

The first step, migrating to fastify v4?

**Comment by zekth** [comment]
@Eomm I'm +1. I can transfer ownership

**Comment by Eomm** [comment]
Go for it, then we will update it 👍🏽 
Easy GitHub activity 🟩🟩🟩🟩🟩🟩 😄

**Comment by zekth** [comment]
Done

**Comment by mcollina** [comment]
@zekth can you add me to npm so I can add it to the org?

**Comment by delvedor** [comment]
Can we close this one?

**Comment by Uzlopak** [comment]
@mcollina 
We never released the plugin.

**Comment by Eomm** [comment]
For reference https://github.com/fastify/fastify-early-hints

**Comment by mcollina** [comment]
> @mcollina 
> 
> We never released the plugin.

why?

**Comment by Uzlopak** [comment]
I dont know. Maybe because @climba03003 suggested to implement node native api call when it is released. It is released with node 18.11.0.

**Comment by climba03003** [comment]
It can be released, but it can improve to use the native one when possible.
Not sure if it should be released as `1.0.0`?

**Comment by climba03003** [comment]
I choose to release in a major to imply it is migrated to organization and a stable release.
https://www.npmjs.com/package/@fastify/early-hints
